src/test/shell/bazel/bazel_cc_code_coverage_test.sh

```diff
@@ -81,38 +86,44 @@ int main(void) {
 }
 EOF
 
-  generate_gcno_files coverage_srcs/a.h coverage_srcs/a.cc coverage_srcs/t.cc
-  generate_instrumented_binary ./coverage_srcs/test coverage_srcs/a.h \
+  generate_instrumented_binary coverage_srcs/test coverage_srcs/a.h \
       coverage_srcs/a.cc coverage_srcs/t.cc
-  generate_gcda_file ./coverage_srcs/test
-}
 
-# Reads the list of arguments provided by the caller (using $@) and uses them
-# to produco .gcno files using g++.
-function generate_gcno_files() {
-  # "-fprofile-arcs -ftest-coverage" tells the compiler to generate coverage
-  # information needed by gcov and include additional code in the object files
-  # for generating the profiling.
-  g++ -fprofile-arcs -ftest-coverage "$@" && return 0
-  fail "Couldn't produce .gcno files for $@"
-  return 1
+  # g++ generates the notes files in the current directory. The documentation
+  # (https://gcc.gnu.org/onlinedocs/gcc/Gcov-Data-Files.html#Gcov-Data-Files)
+  # says they are placed in the same directory as the object file, but they
+  # are not. Therefore we move them in the same directory.
+  mv *.gcno coverage_srcs/
+
+  generate_gcda_file ./coverage_srcs/test
 }
 
 # Reads the list of arguments provided by the caller (using $@) and uses them
-# to produce an instrumented binary using g++.
+# to produce an instrumented binary using g++. At the same time it generates
+# the notes (.gcno) files.
 # - path_to_binary destination of the binary produced by g++
 function generate_instrumented_binary() {
   local path_to_binary="${1}"; shift
-  # "-fprofile-arcs -ftest-coverage" tells the compiler to generate coverage
-  # information needed by gcov and include additional code in the object files
-  # for generating the profiling.
-  g++ -fprofile-arcs -ftest-coverage "$@" -o "$path_to_binary"  && return 0
+  # -fprofile-arcs   Instruments $path_to_binary. During execution the binary
+  #                  records code coverage information.
+  # -ftest-coverage  Produces a notes (.gcno) file that coverage utilities
+  #                  (e.g. gcov, lcov) can use to show a coverage report.
+  # -fprofile-dir    Sets the directory where the profile data (gcda) appears.
```

Comment:
Wow, this is a bit confusing, if you're going to pass the -fprofile-dir then it should be a parameter of this function you can inline the generate_gcda_file here.